### PR TITLE
Simplify RequestParam internal API

### PR DESCRIPTION
**Pull request by brennobemoura**
- Added buildRequest(_:) to RequestParam.
- Implemented SessionConfiguration protocol to set properties to URLSessionConfiguration as internal protocol.
- CombinedParams is not cast over the framework anymore. Added unzip to RequestParam that will check if the self-reference is a CombinedParams type, returning the children if it is, or returning self inside an Array if it isn't.

### CODE REVIEW

**Comment by carson-katri** [comment]
I'm personally not a fan of putting the license in every file... Would you mind reverting that?

**Comment by brennobemoura** [comment]
Alright, no problem

File: Sources/Request/Request/RequestBuilder.swift [hunk 1]
```diff
@@ -10,35 +10,26 @@ import Foundation
 @_functionBuilder
 public struct RequestBuilder {
     public static func buildBlock(_ params: RequestParam...) -> RequestParam {
```
Comment (carson-katri) [inline Sources/Request/Request/RequestBuilder.swift:12]: I think that `unzip` and even `CombinedParams` will be unnecessary if we return an array of `RequestParam`s from the builder:
```suggestion
    public static func buildBlock(_ params: RequestParam...) -> [RequestParam] {
        params
    }
```

File: Sources/Request/Request/RequestBuilder.swift [hunk 2]
```diff
@@ -10,35 +10,26 @@ import Foundation
...
-        if resultParams.filter({ $0.type == .url }).count > 1 {
-            fatalError("You cannot specify more than 1 `Url`")
-        }
-        // Missing Url
-        if resultParams.filter({ $0.type == .url }).count < 1 {
-            fatalError("You must have a `Url`")
-        }
-        return CombinedParams(children: resultParams)
+        CombinedParams(children: params)
     }
```
Comment (carson-katri) [inline Sources/Request/Request/RequestBuilder.swift:14]: I don't think we need `CombinedParams` or `unzip` if we just return an array of params in these builder functions:
```suggestion
    public static func buildBlock(_ params: RequestParam...) -> [RequestParam] {
        params
    }
```

File: Sources/Request/Request/RequestParams/Url/Url.swift [hunk 3]
```diff
@@ -10,30 +10,50 @@ import Foundation
...
-    public init(_ value: String) {
-        self.value = value
+    public init(_ path: String) {
+        self.type = nil
+        self.path = path
     }
     
-    public init(`protocol` type: ProtocolType, url: String) {
-        self.value = "\(type.rawValue)://\(url)"
+    public init(_ type: ProtocolType, url: String) {
```
Comment (carson-katri) [inline Sources/Request/Request/RequestParams/Url/Url.swift:21]: Please bring the `protocol` label back so we don't change the public API.

File: Sources/Request/Request/RequestBuilder.swift [hunk 4]
```diff
@@ -10,35 +10,26 @@ import Foundation
...
-        return param
+        param
     }
 
     public static func buildIf(_ param: RequestParam?) -> RequestParam {
         if let param = param {
             return param
         }
+
         return CombinedParams(children: [])
```
Comment (carson-katri) [inline Sources/Request/Request/RequestBuilder.swift:25]: After removing `CombinedParams`, we can probably add an `EmptyParam` to replace this:
```swift
public struct EmptyParam: RequestParam {
    public func buildParam(_ request: inout URLRequest) {}
}
```

File: Sources/Request/Request/RequestParams/Url/Url.swift [hunk 5]
```diff
@@ -10,30 +10,50 @@ import Foundation
...
+        self.type = type
+        self.path = url
+    }
+
+    public func buildParam(_ request: inout URLRequest) {
+        request.url = URL(string: absoluteString)
+    }
+
+    fileprivate func append(_ string: String) -> Url {
```
Comment (carson-katri) [inline Sources/Request/Request/RequestParams/Url/Url.swift:30]: This may be useful if made `public`.

File: Sources/Request/Request/RequestParams/Query.swift [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    /// Creates the `Query` from `[key:value]` pairs
+    /// - Parameter params: Key-value pairs describing the `Query`
+    public init(_ params: [String: Any?]) {
+        children = params.map {
+            QueryParam($0.key, value: $0.value)
+        }
+    }
+
+    /// Creates the `Query` directly from an array of `QueryParam`s
+    public init(@RequestBuilder _ params: () -> RequestParam) {
```
Comment (carson-katri) [inline Sources/Request/Request/RequestParams/Query.swift:25]: This should have it's own builder, which should probably be done in a separate PR. Please leave the array initializer for compatibility.

File: Sources/Request/Request/RequestParams/QueryParam.swift [hunk 7]
```diff
@@ -9,34 +9,29 @@ import Foundation
...
 /// A key-value pair for a part of the query string
 public struct QueryParam: RequestParam {
-    public var type: RequestParamType = .query
-    public var key: String?
-    public var value: Any?
+    private var key: String
+    private var value: Any?
     
-    public init(_ key: String, value: String) {
+    public init(_ key: String, value: Any?) {
```
Comment (carson-katri) [inline Sources/Request/Request/RequestParams/QueryParam.swift:15]: I think this should still only support `String`, as we don't know how they want it to be encoded into the URL.

File: Sources/Request/Request/RequestParams/AnyParam.swift [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+//  
+//
+//  Created by brennobemoura on 15/11/20.
+//
+
+import Foundation
+
+/// A way to create a custom `RequestParam`
+/// - Important: You will most likely want to use one of the builtin `RequestParam`s, such as: `Url`, `Method`, `Header`, `Query`, or `Body`.
+public struct AnyParam: RequestParam {
```
Comment (carson-katri) [inline Sources/Request/Request/RequestParams/AnyParam.swift:12]: Please mark this as deprecated, since they can now conform to `RequestParam` themselves:
```swift
@available(*, deprecated, message: "`AnyParam` is deprecated. Please conform to the `RequestParam` protocol instead.")
```

File: Sources/Request/Request/RequestParams/RequestConfiguration.swift [hunk 9]
```diff
@@ -0,0 +1,18 @@
+//
+//  File.swift
+//  
+//
+//  Created by brennobemoura on 15/11/20.
+//
+
+import Foundation
+
+protocol SessionConfiguration: RequestParam {
```
Comment (carson-katri) [inline Sources/Request/Request/RequestParams/RequestConfiguration.swift:10]: I think this should be called `SessionParam` to match `RequestParam`. Also please make it `public` so users can make custom ones.

**Review by carson-katri — CHANGES_REQUESTED** [review]

File: Sources/Request/Request/RequestBuilder.swift (hunk 2 above)
Comment (brennobemoura) [reply Sources/Request/Request/RequestBuilder.swift:14]: The functionBuilder `buildBlock(_:)` needs that the input parameter type should be equal to the return type, so I guess that this solution will not work. But, I would like to suggest that we remove the unzip getter and implement the `buildParam(_:)` on `CombinedParams`.

Comment (carson-katri) [reply Sources/Request/Request/RequestBuilder.swift:14]: Making the return type an array of `RequestParam`s actually makes the input and output match, as the input is an array of params. This is a valid function builder:
```swift
@_functionBuilder
struct ArrayBuilder<T> {
    static func buildBlock(_ items: T...) -> [T] {
        items
    }
}
```
I think implementing our function builder like this will simplify it quite a bit. You may need to return a single-item array of `RequestParam`s from the other build functions too.

Comment (brennobemoura) [reply Sources/Request/Request/RequestBuilder.swift:14]: Still, it may work for simple function builders. In case that we have a `buildBlock(_:)` with an `ifBlock(_:)`, this will mess up the types used by the function builder, since an Array of RequestParam doesn't conform to RequestParam.

```
Request {
    Url("https://example.com")
    Query(["message": "Hello World"]
    // Until now RequestBuilder will only execute buildBlock(_:) resulting in [RequestParam] type

    if true {
        Body(["key": "value"])
        // will execute buildBlock(_:) resulting [RequestParam]
    }
    // ifBlocks needs a RequestParam? type not a [RequestParam]?
    // Now, merging the ifBlocks with buildBuild(_:) will be very tricky. 
    // Because now it has to execute a function that will match this buildBock(_: [RequestParam]...)
    // Or this buildBlock(_: [RequestParam], _: RequestParam)
}
```

So, I would like to suggest to keep the CombinedParams and use the unzip getter for a while. I will search more about it and feel free to update this code after merging it.

Comment (carson-katri) [reply Sources/Request/Request/RequestBuilder.swift:14]: Thanks for clarifying. I agree about making `RequestParam` have a `buildParam`

**Review by brennobemoura — COMMENTED** [review]
Updated everything as requested. Only CombinedParams with unzip variable that still remains with no updates.

Without the new SessionParam, all RequestParam were a single constructor that builds the URLRequest object. Now, it can be a SessionParam that needs an instance of URLSessionConfiguration to be built. So, to conforms CombinedParams to all these rules, we may need to implement more functions that will execute all declarative objects, returning only URLSession as the final product.

File: Sources/Request/Request/RequestParams/RequestConfiguration.swift [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+import Foundation
+
+public protocol SessionParam: RequestParam {
+    func buildConfiguration(_ sessionConfiguration: URLSessionConfiguration)
+}
+
+extension SessionParam {
+    public func buildParam(_ request: inout URLRequest) {
+        fatalError("SessionConfiguration shouldn't build URLRequest")
```
Comment (carson-katri) [inline Sources/Request/Request/RequestParams/RequestConfiguration.swift:16]:
```suggestion
        fatalError("SessionParam shouldn't build URLRequest")
```
Please also rename this file to `SessionParam.swift`.

File: Sources/Request/Request/RequestParams/CombinedParams.swift [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+    init(children: [RequestParam]) {
+        self.children = children
+    }
+
+    func buildParam(_ request: inout URLRequest) {
+        children.forEach {
+            $0.buildParam(&request)
+        }
+    }
```
Comment (carson-katri) [inline Sources/Request/Request/RequestParams/CombinedParams.swift:21]: You should be able to remove unzip if you replace that with:
```suggestion
    func buildParam(_ request: inout URLRequest) {
        children
            .sorted { a, _ in (a is Url) }
            .filter { !($0 is SessionParam) }
            .forEach {
                $0.buildParam(&request)
            }
    }
    
    func buildConfiguration(_ configuration: URLSessionConfiguration) {
        children
            .compactMap { $0 as? SessionParam }
            .forEach {
                $0.buildConfiguration(configuration)
            }
    }
```
(and conform to `SessionParam`)

File: Sources/Request/Request/RequestParams/CombinedParams.swift [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+extension RequestParam {
+    var unzip: [RequestParam] {
+        (self as? CombinedParams).map {
+            $0.children
+                .reduce([]) { $0 + $1.unzip }
+        } ?? [self]
+    }
+}
```
Comment (carson-katri) [inline Sources/Request/Request/RequestParams/CombinedParams.swift:31]:
```suggestion
```

File: Sources/Request/Request/Request.swift [hunk 13]
```diff
@@ -43,7 +43,7 @@ public typealias Request = AnyRequest<Data>
 public struct AnyRequest<ResponseType> where ResponseType: Decodable {
     public let combineIdentifier = CombineIdentifier()
 
-    private var params: CombinedParams
+    private var params: [RequestParam]
```
Comment (carson-katri) [inline Sources/Request/Request/Request.swift:46]: If you remove unzip, this should be switched back to:
```suggestion
    private var params: RequestParam
```

File: Sources/Request/Request/Request.swift [hunk 14]
```diff
@@ -117,67 +102,23 @@ public struct AnyRequest<ResponseType> where ResponseType: Decodable {
...
-                if source.contains(.resource) {
-                    configuration.timeoutIntervalForResource = interval
-                }
-            }
+        for config in params where config is SessionParam {
+            (config as? SessionParam)?.buildConfiguration(configuration)
         }
-        
-        
+
```
Comment (carson-katri) [inline Sources/Request/Request/Request.swift:121]: If you make the changes to `CombinedParams`, then this can be simplified to:
```suggestion
        var request = URLRequest(url: URL(string: "https://")!)
        let configuration = URLSessionConfiguration.default
        
        rootParam.buildParam(&request)
        (rootParam as? SessionParam)?.buildConfiguration(configuration)
```

**Review by carson-katri — CHANGES_REQUESTED** [review]
I think `unzip` can be removed and the logic using it in `buildSession` can be moved to it's `buildParam` implementation.

File: Sources/Request/Request/RequestParams/CombinedParams.swift (hunk 11 above)
Comment (brennobemoura) [reply Sources/Request/Request/RequestParams/CombinedParams.swift:21]: I guess we can remove the fatal error on SessionParam since CombinedParams can be it, so it won't be filtered on buildParam.

Comment (carson-katri) [reply Sources/Request/Request/RequestParams/CombinedParams.swift:21]: Maybe leave the fatalError in case of some unseen edge case?

Comment (brennobemoura) [reply Sources/Request/Request/RequestParams/CombinedParams.swift]: Hmmm, I will try using the Never swift type and filter it.

Comment (brennobemoura) [reply Sources/Request/Request/RequestParams/CombinedParams.swift]: Forget using Never, I add one more verification on filter:

```
children
    .sorted { a, _ in (a is Url) }
    .filter { !($0 is SessionParam) || $0 is CombinedParams }
    .forEach {
        $0.buildParam(&request)
    }
```

**Comment by carson-katri** [comment]
This looks good to me, was there anything else you needed to add/change for this PR? If not, I'll merge it now.

**Comment by brennobemoura** [comment]
@carson-katri you can merge it
